myia/abstract/amerge.py

```diff
@@ -247,10 +247,12 @@ def amerge(self, x1: TaggedPossibilities, x2, forced, bp):
 def amerge(self, x1: xtype.TypeMeta, x2, forced, bp):
     if issubclass(x2, x1):
         return x1
-    elif not forced and issubclass(x1, x2):
-        return x2
-    else:
-        raise TypeMismatchError(x1, x2)
+    elif not forced:
+        # Try to infer the smallest super-type containing both types x1 and x2.
+        super_type = xtype.number_super_set(x1, x2)
```

Comment:
You're right, I needed upcast because I have to multiply an integer with a floating value in RNG, and the floating value could be either 16, 32 or 64 bits. But ultimately, I was able to solve the problem by casting integer to the right floating type using a dictionary.